File: bindings/r/R/puddin.R
# ── R/puddin.R ────────────────────────────────────────────────────────────────
#
# R interface to the Puddin shared library.
#
# All functions are vectorised via base R `Vectorize()`.  For large arrays,
# consider using the scalar forms directly inside `vapply()` or `mapply()` for
# maximum control.
#
# All masses are in kilograms; use MSUN to convert from solar masses.
# All angles are in radians; spin magnitudes are dimensionless (0–1).

#' Solar mass in kilograms.
#' @export
MSUN <- 1.988416e30

# ── internal scalar wrappers calling the C ABI ────────────────────────────────

.total_mass_scalar <- function(m1_kg, m2_kg) {
  .C("r_puddin_total_mass",
     m1_kg  = as.double(m1_kg),
     m2_kg  = as.double(m2_kg),
     result = double(1L),
     PACKAGE = "Puddin")$result
}

.mass_ratio_scalar <- function(m1_kg, m2_kg) {
  .C("r_puddin_mass_ratio",
     m1_kg  = as.double(m1_kg),
     m2_kg  = as.double(m2_kg),
     result = double(1L),
     PACKAGE = "Puddin")$result
}

.symmetric_mass_ratio_scalar <- function(m1_kg, m2_kg) {
  .C("r_puddin_symmetric_mass_ratio",
     m1_kg  = as.double(m1_kg),
     m2_kg  = as.double(m2_kg),
     result = double(1L),
     PACKAGE = "Puddin")$result
}

.chirp_mass_scalar <- function(m1_kg, m2_kg) {
  .C("r_puddin_chirp_mass",
     m1_kg  = as.double(m1_kg),
     m2_kg  = as.double(m2_kg),
     result = double(1L),
     PACKAGE = "Puddin")$result
}

.chi_eff_scalar <- function(m1_kg, m2_kg, a1, a2, tilt1, tilt2) {
  .C("r_puddin_chi_eff",
     m1_kg  = as.double(m1_kg),
     m2_kg  = as.double(m2_kg),
     a1     = as.double(a1),
     a2     = as.double(a2),
     tilt1  = as.double(tilt1),
     tilt2  = as.double(tilt2),
     result = double(1L),
     PACKAGE = "Puddin")$result
}

.chi_p_scalar <- function(m1_kg, m2_kg, a1, a2, tilt1, tilt2) {
  .C("r_puddin_chi_p",
     m1_kg  = as.double(m1_kg),
     m2_kg  = as.double(m2_kg),
     a1     = as.double(a1),
     a2     = as.double(a2),
     tilt1  = as.double(tilt1),
     tilt2  = as.double(tilt2),
     result = double(1L),
     PACKAGE = "Puddin")$result
}

# ── public vectorised API ─────────────────────────────────────────────────────

#' Total mass M = m1 + m2
#'
#' @param m1_kg Component mass 1 (kg, numeric vector)
#' @param m2_kg Component mass 2 (kg, numeric vector)
#' @return Total mass in kilograms (numeric vector)
#' @export
#' @examples
#' total_mass(30 * MSUN, 30 * MSUN) / MSUN  # 60
total_mass <- Vectorize(.total_mass_scalar)

#' Mass ratio q = m2 / m1
#'
#' @param m1_kg Primary mass (kg, m1 >= m2, numeric vector)
#' @param m2_kg Secondary mass (kg, numeric vector)
#' @return Dimensionless mass ratio (numeric vector)
#' @export
#' @examples
#' mass_ratio(30 * MSUN, 15 * MSUN)  # 0.5
mass_ratio <- Vectorize(.mass_ratio_scalar)

#' Symmetric mass ratio eta = m1*m2 / M^2
#'
#' @param m1_kg Component mass 1 (kg, numeric vector)
#' @param m2_kg Component mass 2 (kg, numeric vector)
#' @return Dimensionless symmetric mass ratio in (0, 0.25] (numeric vector)
#' @export
#' @examples
#' symmetric_mass_ratio(30 * MSUN, 30 * MSUN)  # 0.25
symmetric_mass_ratio <- Vectorize(.symmetric_mass_ratio_scalar)

#' Chirp mass Mc = (m1*m2)^(3/5) / M^(1/5)
#'
#' @param m1_kg Component mass 1 (kg, numeric vector)
#' @param m2_kg Component mass 2 (kg, numeric vector)
#' @return Chirp mass in kilograms (numeric vector)
#' @export
#' @examples
#' chirp_mass(30 * MSUN, 30 * MSUN) / MSUN  # ~26.1
chirp_mass <- Vectorize(.chirp_mass_scalar)

#' Component masses from chirp mass and mass ratio q
#'
#' @param mc_kg Chirp mass in kilograms (numeric vector)
#' @param q     Mass ratio q = m2/m1 in (0, 1] (numeric vector)
#' @return Named list with elements \code{m1} and \code{m2} in kilograms
#'   (numeric vectors).
#' @export
#' @examples
#' res <- masses_from_chirp_mass_q(chirp_mass(30 * MSUN, 20 * MSUN), 20/30)
#' res$m1 / MSUN  # ~30
#' res$m2 / MSUN  # ~20
masses_from_chirp_mass_q <- function(mc_kg, q) {
  mc_kg <- as.double(mc_kg)
  q_vec <- as.double(q)
  m1 <- mapply(function(mc, qv)
    .C("r_puddin_m1_from_mc_q",
       mc_kg = mc, q = qv, result = double(1L),
       PACKAGE = "Puddin")$result,
    mc_kg, q_vec)
  m2 <- mapply(function(mc, qv)
    .C("r_puddin_m2_from_mc_q",
       mc_kg = mc, q = qv, result = double(1L),
       PACKAGE = "Puddin")$result,
    mc_kg, q_vec)
  list(m1 = unname(m1), m2 = unname(m2))
}

#' Component masses from chirp mass and symmetric mass ratio eta
#'
#' @param mc_kg Chirp mass in kilograms (numeric vector)
#' @param eta   Symmetric mass ratio eta = m1*m2/M^2 in (0, 0.25] (numeric vector)
#' @return Named list with elements \code{m1} and \code{m2} in kilograms
#'   (numeric vectors).
#' @export
#' @examples
#' res <- masses_from_chirp_mass_eta(
#'   chirp_mass(30 * MSUN, 20 * MSUN),
#'   symmetric_mass_ratio(30 * MSUN, 20 * MSUN)
#' )
#' res$m1 / MSUN  # ~30
#' res$m2 / MSUN  # ~20
masses_from_chirp_mass_eta <- function(mc_kg, eta) {
  mc_kg   <- as.double(mc_kg)
  eta_vec <- as.double(eta)
  m1 <- mapply(function(mc, ev)
    .C("r_puddin_m1_from_mc_eta",
       mc_kg = mc, eta = ev, result = double(1L),
       PACKAGE = "Puddin")$result,
    mc_kg, eta_vec)
  m2 <- mapply(function(mc, ev)
    .C("r_puddin_m2_from_mc_eta",
       mc_kg = mc, eta = ev, result = double(1L),
       PACKAGE = "Puddin")$result,
    mc_kg, eta_vec)
  list(m1 = unname(m1), m2 = unname(m2))
}

#' Effective inspiral spin chi_eff
#'
#' @param m1_kg  Component mass 1 (kg, numeric vector)
#' @param m2_kg  Component mass 2 (kg, numeric vector)
#' @param a1     Spin magnitude of body 1, 0-1 (numeric vector)
#' @param a2     Spin magnitude of body 2, 0-1 (numeric vector)
#' @param tilt1  Spin tilt of body 1 in radians, 0-pi (numeric vector)
#' @param tilt2  Spin tilt of body 2 in radians, 0-pi (numeric vector)
#' @return Effective inspiral spin in [-1, 1] (numeric vector)
#' @export
#' @examples
#' chi_eff(30 * MSUN, 30 * MSUN, 0, 0, 0, 0)  # 0
chi_eff <- Vectorize(.chi_eff_scalar)

#' Effective precession spin chi_p
#'
#' Requires m1 >= m2.
#'
#' @param m1_kg  Primary mass (kg, m1 >= m2, numeric vector)
#' @param m2_kg  Secondary mass (kg, numeric vector)
#' @param a1     Spin magnitude of body 1, 0-1 (numeric vector)
#' @param a2     Spin magnitude of body 2, 0-1 (numeric vector)
#' @param tilt1  Spin tilt of body 1 in radians, 0-pi (numeric vector)
#' @param tilt2  Spin tilt of body 2 in radians, 0-pi (numeric vector)
#' @return Effective precession spin in [0, 1] (numeric vector)
#' @export
#' @examples
#' chi_p(30 * MSUN, 30 * MSUN, 0, 0, 0, 0)  # 0
chi_p <- Vectorize(.chi_p_scalar)

# ── binary environment ────────────────────────────────────────────────────────
#
# Groups all compact binary parameter functions under a single domain-organised
# namespace, mirroring `puddin.binary` in Python and `Puddin.Binary` in Julia.
#
# Usage:
#   library(Puddin)
#   Puddin::binary$chirp_mass(30 * MSUN, 30 * MSUN)

#' Compact binary parameter functions for gravitational-wave astronomy.
#'
#' An environment containing all binary system parameter functions, grouped
#' under a single domain namespace.  Mirrors the \code{puddin.binary} module
#' in Python and the \code{Puddin.Binary} module in Julia.
#'
#' Usage:
#' \preformatted{
#' library(Puddin)
#' Puddin::binary$chirp_mass(30 * MSUN, 30 * MSUN) / MSUN  # ~26.1
#' }
#'
#' @export
binary <- new.env(parent = emptyenv())

binary$MSUN                     <- MSUN
binary$total_mass               <- total_mass
binary$mass_ratio               <- mass_ratio
binary$symmetric_mass_ratio     <- symmetric_mass_ratio
binary$chirp_mass               <- chirp_mass
binary$masses_from_chirp_mass_q <- masses_from_chirp_mass_q
binary$masses_from_chirp_mass_eta <- masses_from_chirp_mass_eta
binary$chi_eff                  <- chi_eff
binary$chi_p                    <- chi_p

lockEnvironment(binary, bindings = TRUE)

File: bindings/julia/src/Puddin.jl
"""
    Puddin

Julia interface to the Puddin Rust library for gravitational-wave binary
parameter computations.

All functions accept and return SI values (`Float64`, kilograms for masses,
radians for angles, dimensionless otherwise).

Functions are organised by physics domain into submodules:

- [`Puddin.Binary`](@ref) — compact binary parameter conversions (masses, spins)

All functions are also exported from the top-level `Puddin` module for
convenience, so existing code using `using Puddin; chirp_mass(...)` continues
to work without modification.

Use Julia broadcasting to apply scalar functions over arrays:

```julia
using Puddin

m1 = fill(30.0 * MSUN, 1000)
m2 = fill(30.0 * MSUN, 1000)

mc  = Binary.chirp_mass.(m1, m2)   # via submodule
mc  = chirp_mass.(m1, m2)          # top-level shortcut (backward compat)
eta = symmetric_mass_ratio.(m1, m2)
```
"""
module Puddin

# ── shared library location ───────────────────────────────────────────────────
#
# Development layout (monorepo):
#   cargo build --release -p puddin-julia
#   → <repo-root>/target/release/libpuddin_julia.{so,dylib,dll}
#
# For a registered Julia package the library should instead be supplied by a
# companion JLL package created with BinaryBuilder.jl.

const _REPO_ROOT = joinpath(@__DIR__, "..", "..", "..")
const _LIB = joinpath(_REPO_ROOT, "target", "release", "libpuddin_julia")

# ── Binary submodule ──────────────────────────────────────────────────────────

"""
    Binary

Compact binary system parameter functions for gravitational-wave astronomy.

All functions accept scalar `Float64` values in SI units (kg for mass, radians
for angles).  Vectorisation is handled by Julia broadcasting:

```julia
using Puddin

mc = Binary.chirp_mass.(m1_array, m2_array)
```
"""
module Binary

export MSUN,
       total_mass, mass_ratio, symmetric_mass_ratio, chirp_mass,
       masses_from_chirp_mass_q, masses_from_chirp_mass_eta,
       chi_eff, chi_p

const _REPO_ROOT = joinpath(@__DIR__, "..", "..", "..")
const _LIB = joinpath(_REPO_ROOT, "target", "release", "libpuddin_julia")

"Solar mass in kilograms."
const MSUN::Float64 = 1.988_416e30

"""
    total_mass(m1_kg, m2_kg) -> Float64

Total mass ``M = m_1 + m_2`` in kilograms.
"""
function total_mass(m1_kg::Float64, m2_kg::Float64)::Float64
    ccall((:puddin_total_mass, _LIB), Float64, (Float64, Float64), m1_kg, m2_kg)
end

"""
    mass_ratio(m1_kg, m2_kg) -> Float64

Mass ratio ``q = m_2 / m_1`` (dimensionless, requires ``m_1 \\geq m_2``).
"""
function mass_ratio(m1_kg::Float64, m2_kg::Float64)::Float64
    ccall((:puddin_mass_ratio, _LIB), Float64, (Float64, Float64), m1_kg, m2_kg)
end

"""
    symmetric_mass_ratio(m1_kg, m2_kg) -> Float64

Symmetric mass ratio ``\\eta = m_1 m_2 / M^2 \\in (0, 1/4]`` (dimensionless).
"""
function symmetric_mass_ratio(m1_kg::Float64, m2_kg::Float64)::Float64
    ccall((:puddin_symmetric_mass_ratio, _LIB), Float64, (Float64, Float64), m1_kg, m2_kg)
end

"""
    chirp_mass(m1_kg, m2_kg) -> Float64

Chirp mass ``\\mathcal{M} = (m_1 m_2)^{3/5} / M^{1/5}`` in kilograms.
"""
function chirp_mass(m1_kg::Float64, m2_kg::Float64)::Float64
    ccall((:puddin_chirp_mass, _LIB), Float64, (Float64, Float64), m1_kg, m2_kg)
end

"""
    masses_from_chirp_mass_q(mc_kg, q) -> Tuple{Float64, Float64}

Component masses ``(m_1, m_2)`` in kilograms from chirp mass ``\\mathcal{M}``
(kg) and mass ratio ``q = m_2/m_1 \\in (0, 1]``.

Returns `(m1_kg, m2_kg)` with `m1 ≥ m2`.
"""
function masses_from_chirp_mass_q(mc_kg::Float64, q::Float64)::Tuple{Float64,Float64}
    m1 = ccall((:puddin_m1_from_mc_q, _LIB), Float64, (Float64, Float64), mc_kg, q)
    m2 = ccall((:puddin_m2_from_mc_q, _LIB), Float64, (Float64, Float64), mc_kg, q)
    (m1, m2)
end

"""
    masses_from_chirp_mass_eta(mc_kg, eta) -> Tuple{Float64, Float64}

Component masses ``(m_1, m_2)`` in kilograms from chirp mass ``\\mathcal{M}``
(kg) and symmetric mass ratio ``\\eta \\in (0, 0.25]``.

Returns `(m1_kg, m2_kg)` with `m1 ≥ m2`.
"""
function masses_from_chirp_mass_eta(mc_kg::Float64, eta::Float64)::Tuple{Float64,Float64}
    m1 = ccall((:puddin_m1_from_mc_eta, _LIB), Float64, (Float64, Float64), mc_kg, eta)
    m2 = ccall((:puddin_m2_from_mc_eta, _LIB), Float64, (Float64, Float64), mc_kg, eta)
    (m1, m2)
end

"""
    chi_eff(m1_kg, m2_kg, a1, a2, tilt1, tilt2) -> Float64

Effective inspiral spin ``\\chi_\\mathrm{eff} \\in [-1, 1]``.

- `a1`, `a2`: dimensionless spin magnitudes (0–1)
- `tilt1`, `tilt2`: spin tilt angles in radians (0–π)
"""
function chi_eff(
    m1_kg::Float64, m2_kg::Float64,
    a1::Float64, a2::Float64,
    tilt1::Float64, tilt2::Float64,
)::Float64
    ccall(
        (:puddin_chi_eff, _LIB), Float64,
        (Float64, Float64, Float64, Float64, Float64, Float64),
        m1_kg, m2_kg, a1, a2, tilt1, tilt2,
    )
end

"""
    chi_p(m1_kg, m2_kg, a1, a2, tilt1, tilt2) -> Float64

Effective precession spin ``\\chi_p \\in [0, 1]``.

Requires ``m_1 \\geq m_2``.

- `a1`, `a2`: dimensionless spin magnitudes (0–1)
- `tilt1`, `tilt2`: spin tilt angles in radians (0–π)
"""
function chi_p(
    m1_kg::Float64, m2_kg::Float64,
    a1::Float64, a2::Float64,
    tilt1::Float64, tilt2::Float64,
)::Float64
    ccall(
        (:puddin_chi_p, _LIB), Float64,
        (Float64, Float64, Float64, Float64, Float64, Float64),
        m1_kg, m2_kg, a1, a2, tilt1, tilt2,
    )
end

end # module Binary

# ── top-level re-exports (backward-compatible shortcuts) ──────────────────────

export Binary

export MSUN,
       total_mass, mass_ratio, symmetric_mass_ratio, chirp_mass,
       masses_from_chirp_mass_q, masses_from_chirp_mass_eta,
       chi_eff, chi_p

const MSUN = Binary.MSUN
const total_mass = Binary.total_mass
const mass_ratio = Binary.mass_ratio
const symmetric_mass_ratio = Binary.symmetric_mass_ratio
const chirp_mass = Binary.chirp_mass
const masses_from_chirp_mass_q = Binary.masses_from_chirp_mass_q
const masses_from_chirp_mass_eta = Binary.masses_from_chirp_mass_eta
const chi_eff = Binary.chi_eff
const chi_p = Binary.chi_p

end # module Puddin
